python/tskit/metadata.py

```diff
@@ -0,0 +1,181 @@
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+Classes for metadata decoding, encoding and validation
+"""
+import abc
+import json
+from typing import Any
+
+import jsonschema
+
+import tskit.exceptions as exceptions
+
+
+class AbstractMetadataCodec(metaclass=abc.ABCMeta):
+    """
+    Superclass of all MetadataCodecs
+    """
+
+    def __init__(self) -> None:
+        raise NotImplementedError  # pragma: no cover
+
+    @staticmethod
+    @abc.abstractmethod
+    def encode(obj: Any) -> bytes:
+        raise NotImplementedError  # pragma: no cover
+
+    @staticmethod
+    @abc.abstractmethod
+    def decode(encoded: bytes) -> Any:
+        raise NotImplementedError  # pragma: no cover
+
+
+class JSONCodec(AbstractMetadataCodec):
+    @staticmethod
+    def encode(obj: Any) -> bytes:
+        return json.dumps(obj).encode()
+
+    @staticmethod
+    def decode(encoded: bytes) -> Any:
+        return json.loads(encoded.decode())
+
+
+class NullCodec(AbstractMetadataCodec):
+    @staticmethod
+    def encode(obj: Any) -> bytes:
+        return obj
+
+    @staticmethod
+    def decode(encoded: bytes) -> Any:
+        return encoded
+
+
+class NullValidator:
+    @staticmethod
+    def check_schema(schema: dict) -> None:
+        pass
+
+    def __init__(self, schema: dict) -> None:
+        pass
+
+    def validate(self, obj: Any) -> None:
+        if obj is not None and type(obj) != bytes:
+            raise TypeError(
+                f"expected bytes, found {type(obj)}. "
+                f"To store non-bytes metadata you need to set a "
+                f"metadata_schema on the table"
+            )
+
+
+class MetadataSchema:
+    """
+    Class for validating, encoding and decoding metadata.
+    See the :ref:`metadata <sec_metadata_definition>` section for example usages.
+
+    :param str codec: A string indicating the desired encoding, currently "json"
+        or "null"
+    :param dict schema: A dict containing a valid JSONSchema object. A good guide to
+        creating these is at https://json-schema.org/understanding-json-schema/
+    """
+
+    codecs = {
+        "json": (JSONCodec, jsonschema.validators.Draft7Validator),
+        "null": (NullCodec, NullValidator),
+    }
+
+    def __init__(self, codec: str, schema: dict) -> None:
+        self._schema = schema
+        self._codec = codec
```

Comment:
Shouldn't the ``codec`` be read in as part of the schema? How else are clients to know how to decode the data when they read it in?

Reply:
Confused what you mean here. If you have a metadata_schema string from the C API you should be calling `metadata.parse_metadata_schema`. This is for making a `MetadataSchema` when you are setting it in client table creation code.

Reply:
I'm thinking about the structure of the schema itself, so we have 
```json
 {
            "type": "object",
            "properties": {
                "codec": "json", // A required property in our schema?
                "accession_number": {"type": "number"}, 
                "collection_date": {
                    "name": "Collection date",
                    "description": "Date of sample collection in ISO format",
                    "type": "string", 
                    "pattern": "^([1-9][0-9]{3})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])?$"
                },
            },
            "required": ["accession_number"],
            "additionalProperties": False,
        }
```

So then when client code reads the schema (in whatever language) it knows how to interpret the bytes that are in the ``metadata`` fields.

Reply:
Ahhhh, I see where we are missing each other. A codec applies on a per-table basis, not a per-property basis.

Reply:
The per-property basis is for when using the `struct` codec.